lib/utils/HfbFrameView.hpp

```diff
@@ -0,0 +1,230 @@
+     * don't match an exception is thrown.
+     *
+     * @note To copy the whole frame it is more efficient to use the copying
+     * constructor.
+     *
+     * @param  frame_to_copy  Frame to copy metadata from.
+     * @param  skip_members   Specify a set of data members to *not* copy.
+     *
+     **/
+    void copy_data(HfbFrameView frame_to_copy, const std::set<hfbField>& skip_members);
+
+    /**
+     * @brief Populate metadata.
+     *
+     * @param metadata    Metadata to populate.
+     * @param num_beams   Number of beams.
+     * @param num_subfreq Number of sub-frequencies.
+     *
+     **/
+    static void set_metadata(hfbMetadata* metadata, const uint32_t num_beams,
+                             const uint32_t num_subfreq);
+
+    /**
+     * @brief Populate metadata.
+     *
+     * @param buf         Buffer.
+     * @param index       Index into buffer.
+     * @param num_beams   Number of beams.
+     * @param num_subfreq Number of sub-frequencies.
+     *
+     **/
+    static void set_metadata(Buffer* buf, const uint32_t index, const uint32_t num_beams,
+                             const uint32_t num_subfreq);
+
+    /**
+     * @brief Populate metadata and frame view.
+     *
+     * @param buf            Buffer.
+     * @param index          Index into buffer.
+     * @param num_beams      Number of beams.
+     * @param num_subfreq    Number of sub-frequencies.
+     * @param alloc_metadata Bool to allocate metadata or not.
+     *
+     **/
+    static HfbFrameView create_frame_view(Buffer* buf, const uint32_t index,
+                                          const uint32_t num_beams, const uint32_t num_subfreq,
+                                          bool alloc_metadata = true);
+
+    /**
+     * @brief Read only access to the metadata.
+     * @returns The metadata.
+     **/
+    const hfbMetadata* metadata() const {
+        return _metadata;
+    }
+
+private:
+    // References to the metadata we are viewing
+    hfbMetadata* const _metadata;
+
+    // The calculated layout of the buffer
+    struct_layout<hfbField> buffer_layout;
+
+    // NOTE: these need to be defined in a final public block to ensure that they
+    // are initialised after the above members.
+public:
+    /// The number of beams in the data (read only).
+    const uint32_t& num_beams;
+    /// The number of sub-frequencies in the data (read only).
+    const uint32_t& num_subfreq;
+
+    /// GPS time
+    timespec& time;
+    /// The ICEBoard sequence number
+    int64_t& fpga_seq_num;
+    /// Normalisation fraction
+    float& norm_frac;
+    /// Number of samples integrated
+    uint32_t& num_samples_integrated;
+    /// Number of samples expected
+    uint32_t& num_samples_expected;
+    /// A reference to the frequency ID.
+    uint32_t& freq_id;
+    /// A reference to the dataset ID.
+    dset_id_t& dataset_id;
+
+    /// View of the hyperfine beam data.
+    const gsl::span<float> hfb;
```

Comment:
Is there/should there be a weight dataset for the HFB data? I'm not sure quite how I didn't think about this before, but it seems we could and maybe should, calculate the fast timescale variance of the data (similarly to what's done for the visibility system).

It's not clear that should go in within this commit, but you might want to add space for it into the frame format, and into a field for the frame view, even if it's just initialised to zero for now.